qiskit/synthesis/two_qubit/sqisw_decompose.py

```diff
@@ -0,0 +1,211 @@
+
+    Reference:
+        1. C. Huang et al,
+           *Towards ultra-high fidelity quantum operations:
+           SQiSW gate as a native two-qubit gate (2021)*
+           `arXiv:2105.06074 <https://arxiv.org/abs/2105.06074>`_
+    """
+
+    def __init__(self, euler_basis: Optional[List[str]]):
+        # decomposer for the local single-qubit gates
+        from qiskit.transpiler.passes.optimization.optimize_1q_decomposition import (
+            Optimize1qGatesDecomposition,  # pylint: disable=cyclic-import
+        )
+
+        euler_basis = euler_basis or ["u"]
+        self._decomposer1q = Optimize1qGatesDecomposition(euler_basis)
+
+    def __call__(self, unitary, basis_fidelity=1.0, approximate=True):
+        """Decompose a two-qubit unitary into using the sqrt(iSWAP) gate.
+
+        Args:
+            unitary (Operator or ndarray): a 4x4 unitary to synthesize.
+            basis_fidelity (float): Fidelity of the B gate.
+            approximate (bool): Approximates if basis fidelities are less than 1.0.
+
+        Returns:
+            QuantumCircuit: Synthesized circuit.
+        """
+        u_decomp = TwoQubitWeylDecomposition(Operator(unitary))
+
+        x = u_decomp.a
+        y = u_decomp.b
+        z = u_decomp.c
+
+        A1 = Operator(u_decomp.K1r)
+        A2 = Operator(u_decomp.K1l)
+        B1 = Operator(u_decomp.K2r)
+        B2 = Operator(u_decomp.K2l)
+
+        # in the case that 2 x SQiSW gates are needed
+        if abs(z) <= x - y + _EPS:  # red region
+            V = _interleaving_single_qubit_gates(x, y, z)
+
+            v_decomp = TwoQubitWeylDecomposition(Operator(V))
+
+            D1 = Operator(v_decomp.K1r)
+            D2 = Operator(v_decomp.K1l)
+            E1 = Operator(v_decomp.K2r)
+            E2 = Operator(v_decomp.K2l)
+
+            circuit = QuantumCircuit(2)
+            circuit.append(B1, [0])
+            circuit.append(B2, [1])
+
+            circuit.append(E1.adjoint(), [0])
+            circuit.append(E2.adjoint(), [1])
+            circuit.compose(V, inplace=True)
+            circuit.append(D1.adjoint(), [0])
+            circuit.append(D2.adjoint(), [1])
+
+            circuit.append(A1, [0])
+            circuit.append(A2, [1])
+
+        # in the case that 3 SQiSW gates are needed
+        else:
+            if z < 0:  # blue region
+                # CAN(x, y, z) ~ CAN(x, y, -z)†
+                # so we decompose the adjoint, replace SQiSW with a template in
+                # terms of SQiSW†, then invert the whole thing
+                inverse_decomposition = self.__call__(Operator(unitary).adjoint())
+                inverse_decomposition_with_sqisw_dg = QuantumCircuit(2)
+                for instruction in inverse_decomposition:
+                    if isinstance(instruction.operation, SQiSWGate):
+                        inverse_decomposition_with_sqisw_dg.z(0)
+                        inverse_decomposition_with_sqisw_dg.append(SQiSWGate().inverse(), [0, 1])
```

Comment:
I started trying to fix this as you suggest but thinking about it I think the overhead of adding a new gate (and all that goes along with it like definitions and adding to equivalence library) is not worth it. This problem has to be fixed in Qiskit such that taking the inverse of a gate twice gives the same gate. I opened an issue on this 18 months ago https://github.com/Qiskit/qiskit-terra/issues/6879. This has become a higher priority recently and I think once it gets fixed this automatically gets fixed too. Right now it's not really a problem because if the device supports siswap gate then the generated qasm is correct.